tests/test_tinydb.py

```diff
@@ -92,6 +92,29 @@ def test_insert_multiple_with_ids(db):
                                {'int': 1, 'char': 'c'}]) == [1, 2, 3]
 
 
+def test_insert_invalid_type_raises_error(db):
+    with pytest.raises(ValueError, message='Document is not a Mapping'):
+        db.insert(object())  # object() as an example of a non-mapping-type
+
+
+def test_insert_valid_mapping_type(db):
+    from tinydb.database import Mapping
+
+    class CustomDocument(Mapping):
+        def __init__(self, data):
+            self.data = data
+        def __getitem__(self, key):
+            return self.data[key]
+        def __iter__(self):
+            return iter(self.data)
+        def __len__(self):
+            return len(self.data)
+
+    db.purge()
+    db.insert(CustomDocument({'int': 1, 'char': 'a'}))
```

Comment:
This test is actually failing when JSONStorage is used.

Reply:
That is correct, when using JSONStorage, a custom encoder needs to be passed.

Something like below would work.

````python
class CustomDocument(Mapping):
    def __init__(self, data):
        self.data = data
    def __repr__(self):
        return f'{self.__class__.__name__}({self.data})'
    def __getitem__(self, key):
        return self.data[key]
    def __iter__(self):
        return iter(self.data)
    def __len__(self):
        return len(self.data)


class CustomEncoder(JSONEncoder):
    def default(self, obj):
        return obj.data

db = TinyDB('path/to/to.json', cls=CustomEncoder)
db.insert(CustomDocument({'int': 1, 'char': 'a'}))
assert db.count(where('int') == 1) == 1
````

(The `db` fixture (obviously) uses MemoryStorage.)

What is your suggestion?

Reply:
Oh, thanks for reporting this, @halfdan! It's really unlucky that Python's `json` module doesn't support `collections.abc.Mapping`. There's a [question on StackOverflow](https://stackoverflow.com/q/30062760/997063) regarding the rationale behind this behavior, but no clear answer is given.

As a workaround we could convert all documents to `dict`s during insert using `dict(mapping)` like this:

```python
data[doc_id] = dict(document)
```

We'd have to do that for inserts as well as every other operation that is adding/updating documents to the database.

Coincidentally this would also fix #218.

Reply:
I think this would be the best way to solve it. I've created a PR at #256 to discuss this idea.

Reply:
I too believe that fixing this behavior internally makes sense because not providing a custom converter is an easy mistake to make. I added a few comments to #256.